### PR TITLE
Fix #20508 - Authz plugin enabled with large text/JSON POST payload corrupts body

### LINKED ISSUE #20508: Authz plugin enabled with large text/JSON POST payload corrupts body
While spending time trying to understand why the `docker_cli_authz_unix_test.go` integration test couldn't load **busybox**, I found out that the `drainBody()` implementation is not correct in `pkg/authorization`.

I haven't totally solved it yet, but wanted to get an issue open in case others have ideas and explain what I've seen so far.

First, a bit more explanation about `drainBody`; https://github.com/docker/docker/blob/master/pkg/authorization/authz.go#L124-L141

The code returns `newBody` in two different paths (although I think the middle error path cannot be reached), and that is used to set the actual request body `r.Body`. The problem is that this auto-truncates the body to 1MB, given the new readcloser is based on the buffer created with `maxBodySize`.  To play with how this actually works I created: http://play.golang.org/p/JvBVh_Z5h4

You really will never get an error in the scenario when the sizes are matched on `Peek` and `NewReaderSize`; except for the `EOF` if the data is smaller, which the code handles correctly.

I'm not sure what the right answer is yet, but currently this messes with body data incorrectly and can cause significant problems with REST calls that have binary data.  In a separate PR I'm proposing fixing the content type so that truly binary data doesn't even go through `drainBody()`. See docker/engine-api#98

Secondly, if you modify the [authz integration-cli test](https://github.com/docker/docker/blob/master/integration-cli/docker_cli_authz_unix_test.go#L174-L180) to attempt image load while the plugin is loaded, the Docker REST API panics with a "form has no body"--haven't gotten to the bottom of how `r.Body` is `nil`, but forcing it to not come through `drainBody` (by setting the content type) fixes the problem.

### PR DESCRIPTION
**Pull request by liron-l**
Based on the #20508 discussion, we have changed the following:
1. Send body only if content-type is application/json (based on the
   Docker official daemon REST specification, this is the provided for all
   APIs that requires authorization.
2. Correctly verify that the msg body is smaller than max cap (this was
   the actual bug). Fix includes UT.
3. Minor: Check content length > 0 (it was -1 for load, altough an
   attacker can still modify this)

Signed-off-by: Liron Levin liron@twistlock.com

### CODE REVIEW

File: pkg/authorization/authz.go [hunk 1]
```diff
@@ -54,13 +54,11 @@ type Ctx struct {
...
-	if sendBody(ctx.requestURI, r.Header) {
-		if r.ContentLength < maxBodySize {
-			var err error
-			body, r.Body, err = drainBody(r.Body)
-			if err != nil {
-				return err
-			}
+	if sendBody(ctx.requestURI, r.Header) && r.ContentLength > 0 && r.ContentLength < maxBodySize {
+		var err error
+		body, r.Body, err = drainBody(r.Body)
```
Comment (estesp) [inline pkg/authorization/authz.go]: Will this still truncate `r.Body` in the case where **size(r.Body) > maxBodySize**? Given that is handled by the case of `err == nil` in the reworked `drainBody` function, that return value is the new buffered reader/io.ReadCloser that is based on the `NewReaderSize` returned buffer of `maxBodySize`.  I added a comment in the test as well which may be clearer what I mean..

File: pkg/authorization/authz_unix_test.go [hunk 2]
```diff
@@ -135,6 +137,48 @@ func TestResponseModifier(t *testing.T) {
...
+		t.Fatalf("Error must not be nil '%v'", err)
+	}
+
+	// Large request
+	body, closer, err = drainBody(ioutil.NopCloser(bytes.NewReader([]byte(strings.Repeat("a", maxBodySize*2)))))
+
+	if len(body) != 0 {
+		t.Fatalf("Body must not be copied (request is too lrage), actual length '%d'", len(body))
+	}
+	if closer == nil {
```
Comment (estesp) [inline pkg/authorization/authz_unix_test.go]: Should this also check that `closer` still points to the full data provided to `drainBody()` and not truncated at `maxBodySize`?

File: pkg/authorization/authz_unix_test.go [hunk 3]
```diff
@@ -135,6 +137,48 @@ func TestResponseModifier(t *testing.T) {
...
+	}
+	if err != nil {
+		t.Fatalf("Error must not be nil '%v'", err)
+	}
+
+	// Large request
+	body, closer, err = drainBody(ioutil.NopCloser(bytes.NewReader([]byte(strings.Repeat("a", maxBodySize*2)))))
+
+	if len(body) != 0 {
+		t.Fatalf("Body must not be copied (request is too lrage), actual length '%d'", len(body))
```
Comment (estesp) [inline pkg/authorization/authz_unix_test.go]: minor nit: `lrage` --> `large`

**Comment by estesp** [comment]
Thanks @liron-l. I added some comments/questions inline. I'm no expert on the buffer implementations, so maybe I'm off base, but worth checking.

**Comment by liron-l** [comment]
Thanks for your comments @estesp, all makes sense.
I've added an explicit verification that the new  ```io.ReadCloser``` is always pointing to the start of the body, and that no truncation is done. 
I've also refactor the tests to avoid code duplication.

**Comment by liron-l** [comment]
Thanks for the comments @estesp. I would appreciate if you can review this new iteration.

**Comment by estesp** [comment]
**LGTM** Thanks @liron-l!

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**